### ISSUE TITLE
ClaRa Code generation to many arguments to function call

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#8463]
### Description

For the developer version of ClaRa the model `ClaRa.Components.Furnace.Check.Test_CombustionChamber` fails during code compilation because a function call has not the correct number of arguments, see the [log file](https://libraries.openmodelica.org/branches/master/ClaRa_dev/files/ClaRa_dev_ClaRa.Components.Furnace.Check.Test_CombustionChamber.err).

### Steps to Reproduce

Simulate `ClaRa.Components.Furnace.Check.Test_CombustionChamber` to get
```
ClaRa_dev_ClaRa.Components.Furnace.Check.Test_CombustionChamber_06inz.c:349:548: error: too many arguments to function call, expected 7, have 11
  data->localData[0]->realVars[104] /* combustionChamber.fuelObject.xi_e[6] variable */ = omc_ClaRa_Basics_Media_FuelFunctions_massFraction__i__xi(threadData, tmp9, ((modelica_integer) 6), omc_ClaRa_Basics_Media_FuelTypes_BaseFuel(threadData, data->localData[0]->integerVars[52] /* combustionChamber.fuelObject.fuelModel.N_c CONST */, data->localData[0]->integerVars[53] /* combustionChamber.fuelObject.fuelModel.N_e CONST */, tmp10, tmp11, tmp12, data->localData[0]->integerVars[55] /* combustionChamber.fuelObject.fuelModel.waterIndex CONST */, data->localData[0]->integerVars[54] /* combustionChamber.fuelObject.fuelModel.ashIndex CONST */, tmp13, tmp14, data->localData[0]->realVars[52] /* combustionChamber.fuelObject.fuelModel.T_ref CONST */));
```

### Expected Behavior

SimCode+Templates should generate the correct number of arguments to the function call.

### Version and OS

- OpenModelica Version: OMCompiler v1.19.0-dev.545+g5c10d4c725
- OS: Ubuntu 20.04.3 LTS
- Versions of used Modelica libraries: ClaRa_dev @ c57bace0acbe86a8ae2ccf3e427c60093c5ac95c

**Comment by casella** [comment]
@mahge, would you mind having a look? The library can be cloned from https://www.xrg-server.de/gitlab/Vojacek/clara_om.git

**Comment by AnHeuermann** [comment]
Problem is the `FuelTypes` record has constant binding equations and a function accepts the FuelType as input with a default value.

Here is a minimal working excample:
```modelica
model recordception
  record R
    constant Integer i=5;
  end R;
  
  function myFunc
    input R inR= R();
    output Integer i;
  algorithm
    i := inR.i;
   end myFunc;
 
  Integer i;
  R myR;
equation
  i = myFunc(myR);
end recordception;
```

@perost Could you have a look into this? Using the old frontend somewhat works.

Also changing the above example a bit results in the wrong value for `i`.

```modelica
model recordception
  record R_extended
    extends R(i=3);
  end R_extended;
  
  record R
    constant Integer i=5;
  end R;
  
  function myFunc
    input R inR = R_extended();
    output Integer i;
  algorithm
    i := inR.i;
   end myFunc;
 
  Integer i;
  R_extended myR;
equation
  i = myFunc(myR);
end recordception;
```

**Comment by AnHeuermann** [comment]
Also `R` and `R_extended` are different types. Shouldn't it be needed to have some sort of replacable records in here?

```modelcia
  function myFunc
    input R inR = R_extended();
  end myFunc;
```

**Comment by perost** [comment]
Or wait, there's only one record in that model. So I'm not sure why the OF generates a cast in that case, since it's just casting to the same type that it already is.

**Comment by AnHeuermann** [comment]
I guess we need to change the generated constructor function (in C) to have an input for each record element.

So in
```C
ClaRa_Basics_Media_FuelTypes_BaseFuel omc_ClaRa_Basics_Media_FuelTypes_BaseFuel(threadData_t *threadData, real_array omc_C_LHV, real_array omc_C_cp, modelica_integer omc_waterIndex, modelica_integer omc_ashIndex, real_array omc_defaultComposition, real_array omc_xi_e_waf)
{
  ClaRa_Basics_Media_FuelTypes_BaseFuel tmp1;
  real_array tmp2;
  tmp1._C_LHV = omc_C_LHV;
  tmp1._C_cp = omc_C_cp;
  tmp1._waterIndex = omc_waterIndex;
  tmp1._ashIndex = omc_ashIndex;
  tmp1._defaultComposition = omc_defaultComposition;
  tmp1._xi_e_waf = omc_xi_e_waf;
  tmp1._N_c = ((modelica_integer) 5);
  tmp1._N_e = ((modelica_integer) 5);
  array_alloc_scalar_real_array(&tmp2, 5, (modelica_real)1.0, (modelica_real)1.0, (modelica_real)1.0, (modelica_real)1.0, (modelica_real)1.0);
  tmp1._C_rho = tmp2;
  tmp1._T_ref = 273.15;
  return tmp1;
}
```

we should also have inputs for `modelica_integer N_c`, `modelica_integer N_e`, `modelica_integer T_ref `. The arrays could be an issue, but we'll see.

 The Frontend will replace the values given to the constructor, so the Backend doens't need to change them and we hope that the values are correct when calling the constructor with default values.

**Comment by perost** [comment]
To me it seems like we should do the opposite and not add those arguments to the record constructor call. They're supposed to be protected components in the record constructor, so passing them to the constructor seems like unnecessary work since they're not used in the constructor anyway.

**Comment by kabdelhak** [comment]
Is the user not allowed to call a constructor on that record and change these values? Or are they actually static? If the record is extended you can change these values so i thought you could change them anywhere else also.

**Comment by perost** [comment]
> @AnHeuermann and i talked about it and we think the best solution is if:
> 
>     1. the fronted casts the type of the record to actually fit the correct one if we have `R r = R_extends();` then the constructor has to return the defaults of `R_extends()` but cast to `R` right?
> 
>     2. remove the inputs for protected variables
> 
> 
> Is this correct @perost ?

Maybe, like I said I'm not sure how the casting actually works. I don't think the old frontend actually does anything other than say "this is actually an R", but maybe the code generation handles that. I can try to add the casts and see what happens.

**Comment by perost** [comment]
I added type cast of records in #8941. The model still fails for the same reason, but now there are casts in there for whatever good they may do at least.

I'm not sure if I want to merge that right now or not, maybe at least wait until the library testing works again. But at least you can see how to do it if you want to play around with it.

**Comment by AnHeuermann** [comment]
Next issue from the mwe:

```modelica
model recordception
  record R
    constant Integer i=5;
  end R;
  
  function myFunc
    input R inR = R_extended();
    output Integer i;
  algorithm
    i := inR.i;
   end myFunc;
 
  Integer i;
  R myR(i=7);
equation
  i = myFunc(myR);
end recordception;
```

`i` will be `5` and not `7`, because `myFunc(myR)` is calling the constructor and not using the already existing `myR`.

**Comment by AnHeuermann** [comment]
We need to mark variables that are constant and are equation bound.

For my MWE we can test the binding, but in the ClaRa model we lost the `binding = DAE.EQBOUND()` somewhere:

I get
```
const var UNBOUND: N_c
  constant Integer N_c;
const var UNBOUND: N_e
  constant Integer N_e;
const var UNBOUND: waterIndex
  constant Integer waterIndex;
```

but it should be
```
const var EQBOUND: N_c
  constant Integer N_c;
const var EQBOUND: N_e
  constant Integer N_e;
const var UNBOUND: waterIndex
  constant Integer waterIndex;
```

**Comment by AnHeuermann** [comment]
This specific issue is solved by #9041 and creates a new issue:  #9017

```
record SimulationResult
    resultFile = "",
    simulationOptions = "startTime = 0.0, stopTime = 10.0, numberOfIntervals = 500, tolerance = 1e-06, method = 'dassl', fileNamePrefix = 'ClaRa.Components.Furnace.Check.Test_CombustionChamber', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = ''",
    messages = "Simulation execution failed for model: ClaRa.Components.Furnace.Check.Test_CombustionChamber
src->dim_size[0] != dst->dim_size[0], 4 != 6
assert            | debug   | Failed to copy array. Dimension sizes are not equal and destination array is not flexible.
assert            | info    | simulation terminated by an assertion at initialization
",
```
